prov/netdir/src/netdir_init.c

```diff
@@ -104,22 +104,22 @@ static int ofi_nd_adapter_cb(const ND2_ADAPTER_INFO *adapter, const char *name)
 	info->tx_attr->mode = FI_CONTEXT;
 	info->tx_attr->comp_order = FI_ORDER_STRICT;
 	info->tx_attr->inject_size = (size_t)gl_data.inline_thr;
-	info->tx_attr->size = (size_t)adapter->MaxTransferLength;
+	info->tx_attr->size = (size_t)adapter->MaxInitiatorQueueDepth;
 	/* TODO: if optimization will be needed, we can use adapter->MaxInitiatorSge,
 	 * and use ND SGE to send/write iovecs */
-	info->tx_attr->iov_limit = ND_MSG_IOV_LIMIT;
-	info->tx_attr->rma_iov_limit = ND_MSG_IOV_LIMIT;
+	info->tx_attr->iov_limit = (size_t)adapter->MaxInitiatorSge;
+	info->tx_attr->rma_iov_limit = (size_t)adapter->MaxInitiatorSge;
```

Comment:
Gurgle, the ND code uses ND_MSG_IOV_LIMIT internally to allocate arrays.  And 256 is larger than most apps need and will result in lower performance and more memory consumption than using a more realistic value, like 4...

We can't fix everything here in one pass, so for the tx_attr->iov_limit and rx_attr->iov_limit, please use min(adapter->MaxInitiator, ND_MSG_IOV_LIMIT) in the off chance the adapter max is actually > 256.

For tx_attr->rma_iov_limit the value should be 1.  IB devices do not support RDMA operations to more than 1 remote registered memory region.

Reply:
Hadn't followed through in looking at other used of the ND_MSG_IOV_LIMIT constant. Good catch.